execution-engine/src/native_module_manager.rs

```diff
@@ -0,0 +1,291 @@
+                    continue;
+                }
+
+                let path_unprefixed = PathBuf::from("/").join(path_unprefixed);
+                if path_prefixed.is_dir() {
+                    // Create directory on the VFS with `path_open()`
+                    let prestat = self.native_module_vfs.find_prestat(&path_unprefixed);
+                    if prestat.is_ok() {
+                        let (fd, file_name) = prestat?;
+                        self.native_module_vfs.path_open(
+                            fd,
+                            LookupFlags::empty(),
+                            file_name,
+                            OpenFlags::CREATE,
+                            FileSystem::DEFAULT_RIGHTS,
+                            FileSystem::DEFAULT_RIGHTS,
+                            FdFlags::empty(),
+                        )?;
+                        self.copy_fs_to_vfs(&path_unprefixed)?;
+                    }
+                } else {
+                    // Read file on the kernel fileystem, chunk by chunk
+                    let mut f = File::open(self.native_module_directory.join(&path_prefixed))?;
+                    let mut buf: [u8; 128] = [0; 128];
+
+                    // Copy it to the VFS. First truncate the VFS file first
+                    // then append to it. If the principal doesn't have write
+                    // access, just ignore it
+                    if self.native_module_vfs.write_file_by_absolute_path(&path_unprefixed, vec![], false).is_ok() {
+                        loop {
+                            let n = f.read(&mut buf)?;
+                            if n <= 0 {
+                                break;
+                            }
+                            self.native_module_vfs.write_file_by_absolute_path(&path_unprefixed, buf[..n].to_vec(), true)?;
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Delete native module's filesystem on the kernel filesystem.
+    /// As of now, there is no point in doing this, since native modules have
+    /// read and write access to the entire program's VFS, potentially making
+    /// native module executions stateful. In the future, we might consider
+    /// giving native modules access to only a subset of the program's VFS with
+    /// limited permissions.
+    fn teardown_fs(&self) -> FileSystemResult<()> {
+        remove_dir_all(self.native_module_directory.as_path())?;
+        Ok(())
+    }
+
+    /// Run the native module. The input is passed by the WASM program via the
+    /// native module's special file.
+    pub fn execute(&mut self, input: Vec<u8>) -> FileSystemResult<()> {
+        if self.native_module.is_static() {
+            // Look up native module in the static native modules table
+            let mut nm = STATIC_NATIVE_MODULES
+                .lock()
+                .map_err(|_| ErrNo::Inval)?;
+            let nm = nm
```

Comment:
Minor: can chain together with the previous `let` command.

Reply:
Chaining `let` creates a temporary which is freed while still in use